lib/plugins.py

```diff
@@ -863,10 +863,11 @@ def scan_devices(self):
                     connected[client] = pair
                 else:
                     disconnected_ids.append(pair[1])
-            self.clients = connected
 
-        # Unpair disconnected devices
-        for id_ in disconnected_ids:
-            self.unpair_id(id_)
+            # Unpair disconnected devices
```

Comment:
Yes, this definitely belongs behind the lock.  The original programmers made this type of error all over the place for some reason.  Good catch.

Reply:
Yea, it does belong in the lock, but the reason for the change is that otherwise unpair does not close the device.

Reply:
Any reason why moving the assignment to self.clients down a bit made sense to you?  Really it's "6 of 1, half-dozen of the other".. just curious.

Reply:
The problem is that `unpair_id` depends on `clients` to successfully unpair. When `unpair_id` does not find an xpub using `xpub_by_id`, it calls `_close_client` with the id, which uses `client_lookup` to get the client. `client_lookup` uses `clients` to find the id and if `clients` is assigned before that, it can never actually closes the client. You can try this by raising a `BaseException` in a plugins `get_client` method.

Reply:
Jesus.  Ha. Ok.. good catch then. 👍